File: internal/httpd/listing.go
package httpd

import (
	"bytes"
	"embed"
	"html"
	"html/template"
	"net/url"
	"os"
	"path"
	"sort"
	"strings"
	"time"
)

// entry is one line of a directory listing.
type entry struct {
	// Name carries a trailing slash for a folder, as the Node implementation's
	// cnam does, because that is also the href the browser follows.
	Name    string
	IsFile  bool
	Kind    string
	Size    int64
	ModTime time.Time
}

// isFolder reports whether an entry is a folder. The trailing slash is what
// says so rather than IsFile, which is also false for a socket or a device
// node — those are files as far as the listing is concerned.
func (e entry) isFolder() bool {
	return strings.HasSuffix(e.Name, "/")
}

// bare is the name without the trailing slash a folder carries.
func (e entry) bare() string {
	return strings.TrimSuffix(e.Name, "/")
}

// readDirectory lists a folder: the folders first in the order the filesystem
// reports them, then the files newest first.
//
// This order is the one the legacy dls_directory_reader endpoint answers with,
// so it is left as it was; the browsable page sorts the result itself.
func readDirectory(folder string) ([]entry, error) {
	items, err := os.ReadDir(folder)
	if err != nil {
		return nil, err
	}

	var folders, files []entry
	for _, item := range items {
		info, err := item.Info()
		if err != nil {
			// a name that vanished between the read and the stat is skipped
			// rather than failing the whole listing
			continue
		}
		found := entry{
			Name:    info.Name(),
			IsFile:  info.Mode().IsRegular(),
			Kind:    typeOf(info.Name()).Kind,
			Size:    info.Size(),
			ModTime: info.ModTime(),
		}
		if info.IsDir() {
			found.Name += "/"
			folders = append(folders, found)
			continue
		}
		files = append(files, found)
	}
	sort.SliceStable(files, func(i, j int) bool { return files[i].ModTime.After(files[j].ModTime) })
	return append(folders, files...), nil
}

// The columns a listing can be ordered by. An empty key is the default, which
// is not a column order at all: it groups the folders above the files.
const (
	sortName = "name"
	sortDate = "date"
	sortSize = "size"
	sortType = "type"
)

// sortOrder is what the query string asked the listing to be ordered by.
type sortOrder struct {
	key  string
	desc bool
}

// parseSort reads the order out of a query string. Anything that is not one of
// the four columns falls back to the default rather than being an error: the
// query string is part of a link a user can edit.
func parseSort(query url.Values) sortOrder {
	switch key := query.Get("sort"); key {
	case sortName, sortDate, sortSize, sortType:
		return sortOrder{key: key, desc: query.Get("dir") == "desc"}
	default:
		return sortOrder{}
	}
}

// query is the link that asks for this order.
func (o sortOrder) query() string {
	if o.key == "" {
		return "?"
	}
	if o.desc {
		return "?sort=" + o.key + "&dir=desc"
	}
	return "?sort=" + o.key + "&dir=asc"
}

func (o sortOrder) direction() string {
	if o.desc {
		return "desc"
	}
	return "asc"
}

// nextOrder is the order a click on a column asks for: ascending, then
// descending, then no order at all. The third step is the grouped default the
// listing opens in, so every column has a way to turn itself off rather than
// leaving a sort that can only be swapped for another one.
//
// The script does the same, so a click behaves identically whether or not it
// ran.
func nextOrder(current sortOrder, key string) sortOrder {
	if current.key != key {
		return sortOrder{key: key}
	}
	if !current.desc {
		return sortOrder{key: key, desc: true}
	}
	return sortOrder{}
}

// sortEntries returns the entries in the asked-for order, leaving the slice it
// was given alone. Ties break on the name so that two files of the same size
// keep a fixed order between reloads.
func sortEntries(entries []entry, order sortOrder) []entry {
	sorted := make([]entry, len(entries))
	copy(sorted, entries)

	byName := func(a, b entry) bool {
		return strings.ToLower(a.bare()) < strings.ToLower(b.bare())
	}
	if order.key == "" {
		// the order a file browser opens in: folders above files, both by name
		sort.SliceStable(sorted, func(i, j int) bool {
			if sorted[i].isFolder() != sorted[j].isFolder() {
				return sorted[i].isFolder()
			}
			return byName(sorted[i], sorted[j])
		})
		return sorted
	}

	sort.SliceStable(sorted, func(i, j int) bool {
		a, b := sorted[i], sorted[j]
		var before, after bool
		switch order.key {
		case sortDate:
			before, after = a.ModTime.Before(b.ModTime), b.ModTime.Before(a.ModTime)
		case sortSize:
			before, after = sizeOf(a) < sizeOf(b), sizeOf(b) < sizeOf(a)
		case sortType:
			before, after = kindOf(a) < kindOf(b), kindOf(b) < kindOf(a)
		default:
			before, after = byName(a, b), byName(b, a)
		}
		switch {
		case before:
			return !order.desc
		case after:
			return order.desc
		default:
			return byName(a, b)
		}
	})
	return sorted
}

// sizeOf is what a row sorts by in the size column. A folder has a size on
// disk, but it is the size of the folder itself and says nothing about what is
// in it, so it sorts as nothing.
func sizeOf(item entry) int64 {
	if item.isFolder() {
		return 0
	}
	return item.Size
}

// kindOf is the word the Type column shows.
func kindOf(item entry) string {
	if item.isFolder() {
		return "Folder"
	}
	return item.Kind
}

//go:embed assets
var assets embed.FS

// listingTemplate is parsed once: a page that cannot be built is a bug in the
// embedded template rather than something a request can cause.
var listingTemplate = template.Must(template.ParseFS(assets, "assets/listing.html"))

// loginTemplate is the page that asks for a password. It is a template of its
// own rather than a branch inside the listing, so that "the login page shows no
// folder content" holds because there is nothing there to show, rather than
// because an {{if}} is right.
var loginTemplate = template.Must(template.ParseFS(assets, "assets/login.html"))

// listingStyle and listingScript are inlined into every page. Keeping them out
// of the URL space is deliberate: every path this server answers is a path in
// the served folder, so an asset URL would shadow a real name.
var (
	listingStyle  = template.CSS(mustRead("assets/listing.css"))
	listingScript = template.JS(mustRead("assets/listing.js"))
)

func mustRead(name string) string {
	body, err := assets.ReadFile(name)
	if err != nil {
		panic(err)
	}
	return string(body)
}

// rights is what the page may offer to the account looking at it.
type rights struct {
	Upload       bool
	DeleteFile   bool
	DeleteFolder bool
	Mkdir        bool
	Rename       bool
}

// Any reports whether any per-row action is offered, which is what decides
// whether the actions column is there at all.
func (r rights) Any() bool {
	return r.Rename || r.DeleteFile || r.DeleteFolder
}

type crumb struct {
	Name string
	// Link is empty for the folder being shown, which is not a link to itself.
	Link string
	Sep  bool
}

type column struct {
	Key    string
	Label  string
	Class  string
	Link   string
	Active bool
	Order  string
	Caret  string
}

type listingRow struct {
	Name     string
	Label    string
	Link     string
	IsDir    bool
	Group    string
	Bytes    int64
	Unix     int64
	Kind     string
	Size     string
	Modified string
}

// sessionView is what the page says about who is looking at it.
type sessionView struct {
	// User is the signed in account, empty when nobody is. It is filled only
	// for a request that came in with a session token: someone who
	// authenticated with a header has no session to log out of, and offering
	// the button would be a lie.
	User string
	// CanLogin says an account that may use the form exists and nobody is
	// signed in, which is what puts the Log in button on the page.
	CanLogin bool
	Login    string
	Logout   string
	// Admin is the link to the admin interface, empty for everyone but an
	// admin account signed in with a session while the interface is on.
	Admin string
}

// loginData is the login page.
type loginData struct {
	Path    string
	Action  string
	Cancel  string
	Message string
	Nonce   string
	Style   template.CSS
}

func renderLogin(data loginData) ([]byte, error) {
	var page bytes.Buffer
	if err := loginTemplate.Execute(&page, data); err != nil {
		return nil, err
	}
	return page.Bytes(), nil
}

type listingData struct {
	Path    string
	Folder  string
	Crumbs  []crumb
	Parent  string
	Columns []column
	Entries []listingRow
	Sort    string
	Dir     string
	Rights  rights
	Session sessionView
	Nonce   string
	Style   template.CSS
	Script  template.JS
	// MaxChunkSize is what the client splits a large upload into pieces of; 0
	// means chunked upload is off and a large file is sent as one request, as
	// before.
	MaxChunkSize int64
}

// listingPage renders the browsable directory page.
func listingPage(virtual string, entries []entry, order sortOrder, allowed rights, who sessionView, nonce string, maxChunkSize int64) ([]byte, error) {
	rows := make([]listingRow, 0, len(entries))
	for _, item := range sortEntries(entries, order) {
		row := listingRow{
			Name:     item.bare(),
			Label:    item.Name,
			Link:     (&url.URL{Path: item.Name}).String(),
			IsDir:    item.isFolder(),
			Group:    "0",
			Bytes:    sizeOf(item),
			Unix:     item.ModTime.Unix(),
			Kind:     kindOf(item),
			Size:     readableSize(item.Size),
			Modified: item.ModTime.Format("2006-01-02 15:04"),
		}
		if item.isFolder() {
			row.Group = "1"
			// a folder has no size worth showing, but it does have a date, and
			// without it the date column could not be sorted on
			row.Size = "—"
		}
		rows = append(rows, row)
	}

	data := listingData{
		Path:    virtual,
		Folder:  (&url.URL{Path: virtual}).String(),
		Crumbs:  crumbsOf(virtual),
		Parent:  parentOf(virtual),
		Columns: columnsOf(order),
		Session: who,
		Entries: rows,
		Sort:    order.key,
		Dir:     order.direction(),
		Rights:  allowed,
		Nonce:   nonce,
		Style:   listingStyle,
		Script:  listingScript,

		MaxChunkSize: maxChunkSize,
	}

	var page bytes.Buffer
	if err := listingTemplate.Execute(&page, data); err != nil {
		return nil, err
	}
	return page.Bytes(), nil
}

func columnsOf(order sortOrder) []column {
	defined := []column{
		{Key: sortName, Label: "Name", Class: "c-name"},
		{Key: sortDate, Label: "Modified", Class: "c-mod"},
		{Key: sortType, Label: "Type", Class: "c-kind"},
		{Key: sortSize, Label: "Size", Class: "c-size num"},
	}
	for i := range defined {
		defined[i].Link = nextOrder(order, defined[i].Key).query()
		if defined[i].Key != order.key || order.key == "" {
			continue
		}
		defined[i].Active = true
		if order.desc {
			defined[i].Order, defined[i].Caret = "descending", "▼"
		} else {
			defined[i].Order, defined[i].Caret = "ascending", "▲"
		}
	}
	return defined
}

// crumbsOf breaks the path into the links above it. The first one is the root,
// whose name is the leading slash, so the trail reads as the path itself.
func crumbsOf(virtual string) []crumb {
	root := crumb{Name: "/", Link: "/"}
	parts := strings.Split(strings.Trim(virtual, "/"), "/")
	if len(parts) == 1 && parts[0] == "" {
		root.Link = ""
		return []crumb{root}
	}
	crumbs := make([]crumb, 0, len(parts)+1)
	crumbs = append(crumbs, root)
	walked := "/"
	for i, part := range parts {
		walked += part + "/"
		step := crumb{Name: part, Sep: i < len(parts)-1}
		if i < len(parts)-1 {
			step.Link = (&url.URL{Path: walked}).String()
		}
		crumbs = append(crumbs, step)
	}
	return crumbs
}

// parentOf is the link to the folder above, empty at the root.
func parentOf(virtual string) string {
	cleaned := strings.TrimSuffix(virtual, "/")
	if cleaned == "" || cleaned == "/" {
		return ""
	}
	parent := path.Dir(cleaned)
	if !strings.HasSuffix(parent, "/") {
		parent += "/"
	}
	return parent
}

// readerPage is the answer of the dls_directory_reader endpoint, a plain list
// of links rather than the browsable page. It is what a client that is not a
// browser parses, so it is left exactly as the Node implementation wrote it.
func readerPage(folder string, entries []entry) []byte {
	var page strings.Builder
	page.WriteString("listing directory: " + html.EscapeString(folder) + "\n")
	for _, item := range entries {
		kind := "dir"
		if item.IsFile {
			kind = "file"
		}
		link := (&url.URL{Path: item.Name}).String()
		page.WriteString(`<a href="` + link + `">` + html.EscapeString(item.Name) + `</a>` +
			" - filetype: " + kind + " filesize: " + itoa(item.Size) + "<br/>\n")
	}
	return []byte(page.String())
}

File: internal/httpd/assets/listing.js
// The page is complete without this script: the listing is rendered by the
// server and the sort headers are ordinary links. What is added here is the
// part a link cannot do — sorting without a round trip, filtering, and the
// requests that create, rename, remove and upload.
(function () {
  "use strict";

  var table = document.getElementById("listing");
  if (!table) {
    return;
  }
  var body = table.tBodies[0];
  // folder is the path this page lists, always with a trailing slash and
  // already escaped: every request below appends one escaped segment to it.
  var folder = table.dataset.folder;
  var banner = document.getElementById("banner");
  // maxChunkSize is the largest piece an upload is split into; 0 means
  // chunked upload is off and a large file is sent as one request, as before.
  var maxChunkSize = parseInt(table.dataset.maxChunkSize, 10) || 0;

  function rows() {
    return Array.prototype.slice.call(body.rows).filter(function (row) {
      return row.dataset.name !== undefined;
    });
  }

  function segment(name) {
    return folder + encodeURIComponent(name);
  }

  function say(message, good) {
    banner.textContent = message;
    banner.classList.toggle("good", good === true);
    banner.hidden = false;
  }

  function clear() {
    banner.hidden = true;
  }

  // A refused write is answered with a bare status, so what it meant depends on
  // what was asked for: the same 404 is a folder with something still in it,
  // a name already taken, or a file somebody else removed first.
  var refusals = {
    create: {
      404: "That name is taken, or the folder above it is gone.",
      405: "There is already something with that name."
    },
    rename: {
      404: "The file is gone, or the name you typed is not allowed here.",
      412: "There is already something with that name."
    },
    file: { 404: "That file is already gone." },
    folder: { 404: "The folder still has something in it, or it is already gone." },
    upload: {
      404: "There is already a file with that name.",
      416: "This upload lost sync with the server — try uploading it again."
    }
  };

  function reason(what, status, text) {
    // a session can now run out while the page is open, which is a different
    // thing from never having been allowed
    if (status === 401) {
      return "Your session has ended \u2014 reload the page to log in again.";
    }
    if (status === 403) {
      return "You are not allowed to do that here.";
    }
    if (status === 409) {
      return "The folder above it does not exist.";
    }
    if (status === 413) {
      return text || "The file is larger than this server accepts.";
    }
    var known = refusals[what] || {};
    if (known[status]) {
      return known[status];
    }
    return text || ("The server answered " + status + ".");
  }

  function send(what, method, url, headers) {
    return fetch(url, {
      method: method,
      headers: headers || {},
      credentials: "same-origin"
    }).then(function (res) {
      if (res.ok) {
        return null;
      }
      return res.text().then(function (text) {
        throw new Error(reason(what, res.status, text.trim()));
      });
    });
  }

  function done() {
    window.location.reload();
  }

  function failed(err) {
    say(err.message);
  }

  // --- sorting ---------------------------------------------------------
  //
  // The default order — folders first, then files, both by name — is the one
  // the server rendered, so the header cycles back to it after descending
  // rather than leaving no way to return to it.

  var order = { key: table.dataset.sort || "", dir: table.dataset.dir || "asc" };
  // the columns the server rendered, so that a remembered order naming one it
  // no longer has is thrown away rather than applied to nothing
  var columns = Array.prototype.map.call(table.querySelectorAll("th[data-key]"), function (cell) {
    return cell.dataset.key;
  });
  var remembered = "go-fs.listing.order";

  function value(row, key) {
    if (key === "size") {
      return Number(row.dataset.size);
    }
    if (key === "date") {
      return Number(row.dataset.time);
    }
    if (key === "type") {
      return row.dataset.kind.toLowerCase();
    }
    return row.dataset.name.toLowerCase();
  }

  function compare(a, b, key, sign) {
    var x = value(a, key);
    var y = value(b, key);
    if (x < y) {
      return -sign;
    }
    if (x > y) {
      return sign;
    }
    // a stable tie break, so two files of the same size never swap places
    return a.dataset.name.toLowerCase() < b.dataset.name.toLowerCase() ? -1 : 1;
  }

  function arrange() {
    var list = rows();
    var sign = order.dir === "desc" ? -1 : 1;
    if (order.key === "") {
      list.sort(function (a, b) {
        var group = Number(b.dataset.dir) - Number(a.dataset.dir);
        return group !== 0 ? group : compare(a, b, "name", 1);
      });
    } else {
      var key = order.key;
      list.sort(function (a, b) {
        return compare(a, b, key, sign);
      });
    }
    list.forEach(function (row) {
      body.appendChild(row);
    });

    Array.prototype.forEach.call(table.tHead.rows[0].cells, function (cell) {
      var key = cell.dataset.key;
      if (key === undefined) {
        return;
      }
      if (key === order.key && order.key !== "") {
        cell.setAttribute("aria-sort", order.dir === "desc" ? "descending" : "ascending");
        cell.querySelector(".caret").textContent = order.dir === "desc" ? "▼" : "▲";
      } else {
        cell.removeAttribute("aria-sort");
        cell.querySelector(".caret").textContent = "";
      }
    });
  }

  // next is the state a click on a column moves to: ascending, descending, then
  // no order at all, which is the grouped default the page was rendered in.
  function next(key) {
    if (order.key !== key) {
      return { key: key, dir: "asc" };
    }
    if (order.dir === "asc") {
      return { key: key, dir: "desc" };
    }
    return { key: "", dir: "asc" };
  }

  function retarget() {
    Array.prototype.forEach.call(table.querySelectorAll("th[data-key] a"), function (link) {
      var state = next(link.parentNode.dataset.key);
      link.setAttribute("href", query(state));
    });
  }

  function query(state) {
    if (state.key === "") {
      return "?";
    }
    return "?sort=" + state.key + "&dir=" + state.dir;
  }

  // show writes the order into the address bar, so that a reload, a copied link
  // and what is on screen all say the same thing.
  function show() {
    var url = window.location.pathname;
    if (order.key !== "") {
      url += query(order);
    }
    window.history.replaceState(null, "", url);
  }

  // The order is remembered for the server rather than for one folder: somebody
  // who sorted by size wants the folder they open next sorted by size too.
  // Storage can be switched off or full, which costs the preference and nothing
  // else, so every use of it is allowed to fail quietly.
  function keep(state) {
    try {
      if (state.key === "") {
        window.localStorage.removeItem(remembered);
      } else {
        window.localStorage.setItem(remembered, state.key + ":" + state.dir);
      }
    } catch (ignored) {
      return;
    }
  }

  function recall() {
    var stored = null;
    try {
      stored = window.localStorage.getItem(remembered);
    } catch (ignored) {
      return null;
    }
    if (!stored) {
      return null;
    }
    var parts = stored.split(":");
    if (columns.indexOf(parts[0]) === -1 || (parts[1] !== "asc" && parts[1] !== "desc")) {
      // written by an older version, or by hand
      keep({ key: "" });
      return null;
    }
    return { key: parts[0], dir: parts[1] };
  }

  Array.prototype.forEach.call(table.querySelectorAll("th[data-key] a"), function (link) {
    link.addEventListener("click", function (event) {
      event.preventDefault();
      order = next(link.parentNode.dataset.key);
      arrange();
      retarget();
      keep(order);
      show();
    });
  });

  // A query string is somebody asking for an order outright — a link they were
  // sent, or a reload — so it wins and becomes the preference. With none, the
  // preference is what decides, which is how an order survives walking into
  // another folder: the links between folders are relative and carry nothing.
  if (new URLSearchParams(window.location.search).has("sort")) {
    keep(order);
  } else {
    var restored = recall();
    if (restored) {
      order = restored;
      arrange();
      retarget();
      show();
    }
  }

  // --- filtering -------------------------------------------------------

  var filter = document.getElementById("filter");
  if (filter) {
    filter.addEventListener("input", function () {
      var needle = filter.value.trim().toLowerCase();
      rows().forEach(function (row) {
        row.classList.toggle("gone",
          needle !== "" && row.dataset.name.toLowerCase().indexOf(needle) === -1);
      });
    });
  }

  // --- dialogs ---------------------------------------------------------

  var asking = document.getElementById("prompt");
  var checking = document.getElementById("confirm");
  var onAccept = null;
  var onAgree = null;

  function ask(title, detail, current, label, then) {
    asking.querySelector("h2").textContent = title;
    asking.querySelector("p").textContent = detail;
    asking.querySelector(".go").textContent = label;
    var field = asking.querySelector("input");
    field.value = current;
    onAccept = function () {
      var typed = field.value.trim();
      if (typed === "" || typed === current) {
        return;
      }
      then(typed);
    };
    asking.showModal();
    field.focus();
    var stem = current.lastIndexOf(".");
    field.setSelectionRange(0, stem > 0 ? stem : current.length);
  }

  // The submit event rather than close: a dialog closed by its own form does
  // not reliably report a close, and submit says which button was pressed
  // while the answer is still in front of it. An implicit submit — Enter in the
  // field — names no button, and the markup puts the acting one first so that
  // it is the one the browser would have picked anyway.
  function answered(dialog, taken) {
    dialog.querySelector("form").addEventListener("submit", function (event) {
      var pending = taken();
      if (event.submitter && event.submitter.value === "cancel") {
        return;
      }
      if (pending) {
        pending();
      }
    });
  }

  answered(asking, function () {
    var pending = onAccept;
    onAccept = null;
    return pending;
  });

  answered(checking, function () {
    var pending = onAgree;
    onAgree = null;
    return pending;
  });

  // --- actions ---------------------------------------------------------

  var makeFolder = document.getElementById("new-folder");
  if (makeFolder) {
    makeFolder.addEventListener("click", function () {
      clear();
      ask("New folder", "It is created in " + decodeURI(folder) + ".", "", "Create", function (name) {
        send("create", "MKCOL", segment(name) + "/").then(done).catch(failed);
      });
    });
  }

  body.addEventListener("click", function (event) {
    var button = event.target.closest("button[data-do]");
    if (!button) {
      return;
    }
    clear();
    var row = button.closest("tr");
    var name = row.dataset.name;
    var isFolder = row.dataset.dir === "1";
    if (button.dataset.do === "rename") {
      ask("Rename", decodeURI(folder) + name, name, "Rename", function (typed) {
        send("rename", "MOVE", segment(name), { Destination: segment(typed) }).then(done).catch(failed);
      });
      return;
    }
    checking.querySelector("h2").textContent = isFolder ? "Delete folder" : "Delete file";
    checking.querySelector("p").textContent = isFolder
      ? "Delete " + decodeURI(folder) + name + "? Only an empty folder can be removed."
      : "Delete " + decodeURI(folder) + name + "? This cannot be undone.";
    onAgree = function () {
      send(isFolder ? "folder" : "file", "DELETE", segment(name)).then(done).catch(failed);
    };
    checking.showModal();
  });

  // --- uploading -------------------------------------------------------
  //
  // XMLHttpRequest rather than fetch, because only it reports how far an
  // upload has got.

  var drop = document.getElementById("drop");
  if (!drop) {
    return;
  }
  var picker = document.getElementById("picker");
  var queue = document.getElementById("queue");

  document.getElementById("upload").addEventListener("click", function () {
    picker.click();
  });
  picker.addEventListener("change", function () {
    accept(picker.files);
    picker.value = "";
  });

  ["dragenter", "dragover"].forEach(function (name) {
    document.addEventListener(name, function (event) {
      event.preventDefault();
      drop.classList.add("over");
    });
  });
  document.addEventListener("dragleave", function (event) {
    if (event.target === drop || event.relatedTarget === null) {
      drop.classList.remove("over");
    }
  });
  document.addEventListener("drop", function (event) {
    event.preventDefault();
    drop.classList.remove("over");
    if (event.dataTransfer && event.dataTransfer.files.length) {
      accept(event.dataTransfer.files);
    }
  });

  function accept(files) {
    clear();
    var pending = Array.prototype.slice.call(files);
    if (!pending.length) {
      return;
    }
    queue.hidden = false;
    // every file gets its own reason: dropping ten at once and being told only
    // why the last one failed says nothing about the other nine
    var problems = [];
    function step() {
      if (!pending.length) {
        if (problems.length === 0) {
          done();
          return;
        }
        // no reload, because it would take the reasons with it
        say(problems.join("\n"));
        return;
      }
      put(pending.shift(), function (problem) {
        if (problem) {
          problems.push(problem);
        }
        step();
      });
    }
    step();
  }

  function put(file, then) {
    var line = document.createElement("li");
    var label = document.createElement("span");
    label.className = "what";
    label.textContent = file.name;
    var track = document.createElement("span");
    track.className = "track";
    var fill = document.createElement("span");
    fill.className = "fill";
    track.appendChild(fill);
    var state = document.createElement("span");
    state.className = "state";
    state.textContent = "0%";
    line.appendChild(label);
    line.appendChild(track);
    line.appendChild(state);
    queue.appendChild(line);

    function progress(fraction) {
      var percent = Math.round(fraction * 100);
      fill.style.width = percent + "%";
      state.textContent = percent + "%";
    }
    function succeed() {
      fill.style.width = "100%";
      state.textContent = "done";
      then(null);
    }
    function fail(problem) {
      line.classList.add("failed");
      fill.style.width = "100%";
      state.textContent = "failed";
      line.title = problem;
      then(problem);
    }

    if (!maxChunkSize || file.size <= maxChunkSize) {
      putWhole(file, progress, succeed, fail);
      return;
    }
    putChunked(file, progress, succeed, fail);
  }

  // putWhole sends a file as a single request, exactly as every upload was
  // sent before chunking existed: still the path for anything at or under
  // maxChunkSize, and for every upload when chunking is off.
  function putWhole(file, progress, succeed, fail) {
    var request = new XMLHttpRequest();
    request.open("PUT", segment(file.name));
    request.setRequestHeader("Content-Type", "application/octet-stream");
    request.withCredentials = true;
    request.upload.addEventListener("progress", function (event) {
      if (event.lengthComputable) {
        progress(event.loaded / event.total);
      }
    });
    request.addEventListener("load", function () {
      if (request.status >= 200 && request.status < 300) {
        succeed();
        return;
      }
      fail(file.name + ": " + reason("upload", request.status, request.responseText.trim()));
    });
    request.addEventListener("error", function () {
      fail(file.name + ": the upload could not be sent.");
    });
    request.send(file);
  }

  // chunkRetries is how many times the current chunk of a chunked upload is
  // retried before the whole file is reported as failed. A chunk that lands
  // stays landed on the server (its staging file only grows, never shrinks
  // below a completed chunk's end), so retrying is always safe: it never
  // restarts the file from the beginning.
  var chunkRetries = 3;

  // putChunked sends a file as consecutive Content-Range PUTs of at most
  // maxChunkSize each, to the same URL a whole-file PUT uses. The server's
  // staging file is the only record of how far the upload has gotten; a 416
  // means the server's idea of that offset differs from this one (an earlier
  // attempt's chunk landing after this page stopped waiting for it, most
  // likely), and is answered by resuming from the offset the response names,
  // once, rather than treating a resync as a failed upload.
  function putChunked(file, progress, succeed, fail) {
    var sent = 0;
    var retries = 0;
    var resynced = false;

    function sendFrom(start) {
      var end = Math.min(start + maxChunkSize, file.size);

      function retry(problem) {
        if (retries < chunkRetries) {
          retries++;
          sendFrom(start);
          return;
        }
        fail(problem);
      }

      var request = new XMLHttpRequest();
      request.open("PUT", segment(file.name));
      request.setRequestHeader("Content-Type", "application/octet-stream");
      request.setRequestHeader("Content-Range",
        "bytes " + start + "-" + (end - 1) + "/" + file.size);
      request.withCredentials = true;
      request.upload.addEventListener("progress", function (event) {
        if (event.lengthComputable) {
          progress((sent + event.loaded) / file.size);
        }
      });
      request.addEventListener("load", function () {
        if (request.status >= 200 && request.status < 300) {
          sent = end;
          retries = 0;
          progress(sent / file.size);
          if (sent >= file.size) {
            succeed();
          } else {
            sendFrom(sent);
          }
          return;
        }
        if (request.status === 416 && !resynced) {
          var have = resumeOffset(request.getResponseHeader("Content-Range"));
          if (have !== null && have <= file.size) {
            resynced = true;
            sent = have;
            sendFrom(have);
            return;
          }
        }
        retry(file.name + ": " + reason("upload", request.status, request.responseText.trim()));
      });
      request.addEventListener("error", function () {
        retry(file.name + ": the upload could not be sent.");
      });
      request.send(file.slice(start, end));
    }

    sendFrom(0);
  }

  // resumeOffset reads the size a 416 names in its own Content-Range, the
  // "bytes */<size>" form the server answers an out-of-sync chunk with.
  function resumeOffset(header) {
    var match = /^bytes \*\/(\d+)$/.exec(header || "");
    return match ? parseInt(match[1], 10) : null;
  }
})();
